### PR TITLE
feat(code-template): Add Code Template System and Tool Calling Infrastructure

### PR DESCRIPTION
**Pull request by thecodacus**
# Add Code Template System and Tool Calling Infrastructure

## Overview
This PR introduces two major architectural enhancements:
1. A code template system for project scaffolding
2. A foundational infrastructure for AI tool calling with an agent-based architecture

These changes enable more sophisticated project initialization and lay the groundwork for AI-driven tooling capabilities.

### 1. Template System Implementation
- Added project template import functionality
- Implemented GitHub-based template fetching
- Added template configuration and management
- Introduced template ignore patterns for selective imports
- Added support for template-specific prompts

### 2. Tool Calling Infrastructure
- Implemented base agent architecture for tool execution
- Added tool calling parser and execution pipeline
- Integrated tool configuration management
- Added support for agent-based routing and coordination
- Implemented template variables system for dynamic prompting

### 3. Agent System
- Added `AgentPromptGenerator` for structured prompt generation
- Implemented `AgentOutputParser` for processing tool calls
- Added coordinator agent for query routing
- Implemented project scaffolding agent
- Added support for agent registration and management

### 4. UI Enhancements
- Added tool configuration toggle in chat interface
- Implemented tool execution status display
- Added template selection interface
- Enhanced message parsing for tool responses
- Added support for tool calling visualization

### Agent Architecture
- Base agent class with configurable tools and parameters
- XML-based tool call protocol for structured communication
- Support for template variables and dynamic prompt generation
- Coordinator agent for intelligent query routing

### Template System
- GitHub API integration for template fetching
- Support for .bolt configuration directory
- Template ignore patterns for selective imports
- Custom prompt support for templates

### Tool Execution Pipeline
1. Tool call parsing and validation
2. Agent selection and routing
3. Parameter validation
4. Tool execution
5. Response formatting and streaming

## Testing
- Tested template import functionality
- Verified tool calling pipeline
- Tested agent routing system
- Verified template ignore patterns
- Tested tool configuration persistence

## Migration Impact
- No breaking changes for existing chat functionality
- Tool calling requires explicit opt-in via toggle
- Template system is additive to existing functionality
- Existing message parsing remains unchanged

## Future Enhancements
- Add support for more complex tool chains
- Enhance template management system
- Add more specialized agents
- Improve tool execution visualization
- Add support for concurrent tool execution

# Preview

https://github.com/user-attachments/assets/ba3c180c-625b-4d5d-87d1-2580ed6ff5a0

### CODE REVIEW

**Comment by thecodacus** [comment]
@wonderwhy-er , sorry but I tried to make the changes as little as possible 😅

**Comment by thecodacus** [comment]
Also this approach is text based, to it can be used by any llm, as long as it follows the structured output and does not get confused. that require prompt tuning which can be done in separate task

we can now add agents for github project import or commit, using this framework

**Comment by wonderwhy-er** [comment]
Ok, will need to take a deeper look in to how it works

**Comment by thecodacus** [comment]
this is an example of git import using the tooling framework 

https://github.com/user-attachments/assets/6ab045c6-9614-4c41-9410-7263ff10affe

**Comment by wonderwhy-er** [comment]
Ahh its still big, will take time to fully review...

**Comment by Prathamesh-Shanbhag** [comment]
This was an awesome addition - Now all I look forward to is the ability to have attachments as reference to the prompts and one click deploy to vercel or other similar platforms!!

**Comment by wonderwhy-er** [comment]
I did not even notice it had github integration in some form.
If I understand it correctly its for local use only, or one who hosts would need to add his key in there.

Considering its not oauth based all users of hosted app will have 5k requests per hour
https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28

We still need to get to oauth variant to work at some point.

**Comment by thecodacus** [comment]
> I did not even notice it had github integration in some form. If I understand it correctly its for local use only, or one who hosts would need to add his key in there.
> 
> Considering its not oauth based all users of hosted app will have 5k requests per hour https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28
> 
> We still need to get to oauth variant to work at some point.

Yes, for code template system to be downloaded. we will need the key, but the api also has some allowance without any key as well. also we can modify the agent code to use the tarball download. that will not require any oauth or api key at all

for the github project import demo that i showed in the threaded video, was just to show case what we can do with this system, this PR does not contain the github agent included. that is something I just made with this system quickly to showcase

**Comment by adriano2023666** [comment]
I am currently facing a NotFoundError in the application, and I need assistance in resolving it. The error message is as follows:

php
Copiar código
NotFoundError: Failed to execute 'insertBefore' on 'Node': The node before which the new node is to be inserted is not a child of this node.
    at insertBefore (http://localhost:5173/node_modules/.vite/deps/chunk-TLFX46WQ.js?v=b44fa53c:15324:26)
    at insertOrAppendPlacementNode (http://localhost:5173/node_modules/.vite/deps/chunk-TLFX46WQ.js?v=b44fa53c:24310:15)
    at commitPlacement (http://localhost:5173/node_modules/.vite/deps/chunk-TLFX46WQ.js?v=b44fa53c:24267:15)
    at commitReconciliationEffects (http://localhost:5173/node_modules/.vite/deps/chunk-TLFX46WQ.js?v=b44fa53c:24784:15)
    ...
The issue seems to be related to the insertBefore method in the DOM, where the new node cannot be inserted because the specified reference node is not a child of the parent node. This is causing the application to behave unexpectedly, and it's affecting the UI by preventing certain elements from being rendered or placed correctly.

Could anyone help me understand what might be causing this error and guide me on how to resolve
bandicam 2024-11-23 17-54-48-349

File: app/components/chat/APIKeyManager.tsx [hunk 1]
```diff
@@ -44,7 +44,7 @@ export const APIKeyManager: React.FC<APIKeyManagerProps> = ({
       ) : (
         <>
           <span className="flex-1 text-sm text-bolt-elements-textPrimary">
-            {apiKey ? '••••••••' : 'Not set (will still work if set in .env file)'}
+            {apiKey ? '••••••••' : 'Not set (works via .env)'}
```
[the remark this answers is not in the document]
Comment (thecodacus) [reply app/components/chat/APIKeyManager.tsx:47]: this was dont to make some room for the tool manager toggle button, not essential, but yes can be done in separate PR

File: app/components/chat/Chat.client.tsx [hunk 2]
```diff
@@ -104,20 +110,34 @@ export const ChatImpl = memo(({ initialMessages, storeMessageHistory }: ChatProp
     },
     initialMessages,
   });
+  const waitForLoading = useWaitForLoading(isLoading);
```
[the remark this answers is not in the document]
Comment (thecodacus) [reply app/components/chat/Chat.client.tsx:113]: not really, this is needed cause, once the tool call is done I need a way to push the result of the tool call into the conversation  to provide it as context for the llm ( its hidden in the UI ). 

but you can only push anything once the loading is done. so this was implemented for that reason

**Pull request closed without merging**